### PR TITLE
Tutorial state check and autofill

### PR DESCRIPTION
**Pull request by iGio90**
Please, delete my other PR as it doesn't really have nothing to do with their ToS. The call is actually doing the purpose we would reach (let a fresh account receive valid response). I want to externalize the class so it could be customized but tomorrow i'll have a flight and i won't be at pc. Im gonna update this in 2 days but in the meantime you can use this. You can call it safely at every login I guess since the tutorial state can't be downgraded

### CODE REVIEW

**Comment by MrPat** [comment]
I tried with this code and can confirm that it works with a fresh account, but I did have to login again after calling your function in order to get server responses. Nice work!

**Comment by s7092910** [comment]
It works, but heads up you might want to add a Thread.sleep(50) between the get player and the mark complete calls. Or else you have a chance of failing the second call.

**Comment by iGio90** [comment]
Yes! but i got some fabric reports from almost 3k users experiencing this

Fatal Exception: com.pokegoapi.exceptions.AsyncPokemonGoException: Unknown exception occurred. 
       at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:45)
       at com.pokegoapi.main.RequestHandler.sendServerRequests(RequestHandler.java:155)
       at com.pokegoapi.api.PokemonGo.setTutorialState(PokemonGo.java:222)
       at com.pokemesh.helpers.LoginHelper$1.run(LoginHelper.java:120)
       at java.lang.Thread.run(Thread.java:818)
Caused by java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null

The requests has actually sent together. I checked how is the behavior of the official client and the calls are sent almost together. I'm gonna try with the sleep and also to emulate all the other requests that runs at login

**Comment by iGio90** [comment]
I also want to try to make it a stock call when the login is fired. As i said im actually spamming it on a lvl 15 account without issue, so i think it can be safely added to the login flow

**Comment by s7092910** [comment]
I might also suggest that you figure out a way to skip the mark complete if a user has already done it.

**Comment by iGio90** [comment]
@s7092910 i also tought at this! I think i can do it but i have to check if these infos are sent with the stock login flow requests (I think so?).

**Comment by iGio90** [comment]
They are here. I can do it:

[<-] Request for GetPlayer  {}

[->] Response for GetPlayer  {
    "success": true,
    "player_data": {
        "creation_timestamp_ms": "1467910353904",
        "username": "GioBrambilla",
        "team": "BLUE",
        "tutorial_state": [
            "LEGAL_SCREEN",
            "AVATAR_SELECTION",
            "POKEMON_CAPTURE",
            "NAME_SELECTION",
            "FIRST_TIME_EXPERIENCE_COMPLETE"
        ],
        "avatar": {
            "hair": 2,
            "shirt": 1,
            "pants": 1,
            "hat": 2,
            "shoes": 1,
            "eyes": 2,
            "backpack": 2
        },
        "max_pokemon_storage": 250,
        "max_item_storage": 350,
        "daily_bonus": {},
        "equipped_badge": {},
        "contact_settings": {},
        "currencies": [
            {
                "name": "POKECOIN"
            },
            {
                "name": "STARDUST",
                "amount": 6265
            }
        ]
    },
    "$unknownFields": []
}

**Comment by iGio90** [comment]
Ok guys, please check it up! (Still WIP, i need to update the list in our class and make the enableAccount private maybe since it will be used only at login and it's internal)

@MrPat @s7092910

File: library/src/main/java/com/pokegoapi/api/player/PlayerProfile.java [hunk 1]
```diff
@@ -305,4 +293,34 @@ public Stats getStats()
...
+
+	/**
+	 * Gets tutorial states
+	 *
+	 * @return TutorialState object
+	 * @throws LoginFailedException when the auth is invalid
+	 * @throws RemoteServerException when the server is down/having issues
+	 */
+	public TutorialState getTutorialState()
+			throws LoginFailedException, RemoteServerException {
```
Comment (FabianTerhorst) [inline library/src/main/java/com/pokegoapi/api/player/PlayerProfile.java]: Unneeded throws?

File: library/src/main/java/com/pokegoapi/util/Signature.java [hunk 2]
```diff
@@ -79,9 +79,8 @@ public static void setSignature(PokemonGo api, RequestEnvelopeOuterClass.Request
 	}
 
 
-	private static int getLocationHash1(PokemonGo api, byte[] authTicket,
-										RequestEnvelopeOuterClass.RequestEnvelope.Builder builder) {
-		XXHashFactory factory = XXHashFactory.fastestInstance();
+	private static int getLocationHash1(PokemonGo api, byte[] authTicket, RequestEnvelopeOuterClass.RequestEnvelope.Builder builder) {
+		XXHashFactory factory = XXHashFactory.safeInstance();
```
Comment (FabianTerhorst) [inline library/src/main/java/com/pokegoapi/util/Signature.java]: Wrong pr

Comment (iGio90) [inline library/src/main/java/com/pokegoapi/util/Signature.java]: yeah, i merged another PR on my branch! im gonna clean it up once i finish it and also strip out the un-needed throws! can you just check if it works meanwhile? im waiting for some score

File: library/src/main/java/com/pokegoapi/api/PokemonGo.java [hunk 3]
```diff
@@ -18,9 +18,13 @@
 import POGOProtos.Networking.Envelopes.RequestEnvelopeOuterClass.RequestEnvelope.AuthInfo;
 import POGOProtos.Networking.Envelopes.SignatureOuterClass;
 import POGOProtos.Networking.Envelopes.Unknown6OuterClass;
+import POGOProtos.Networking.Requests.RequestTypeOuterClass.RequestType;
```
Comment (FabianTerhorst) [inline library/src/main/java/com/pokegoapi/api/PokemonGo.java]: Are you sure this imporrs are needed, because you doesn't changed the code in this file.

File: library/src/main/java/com/pokegoapi/api/player/PlayerProfile.java [hunk 4]
```diff
@@ -305,4 +293,34 @@ public Stats getStats()
...
+				.setSendMarketingEmails(false)
+				.setSendPushNotifications(false);
+
+		ServerRequest serverRequest = new ServerRequest(RequestType.MARK_TUTORIAL_COMPLETE, tutorialBuilder.build());
+		api.getRequestHandler().sendServerRequests(serverRequest);
+
+		MarkTutorialCompleteResponse response;
+		try {
+			response = MarkTutorialCompleteResponse.parseFrom(serverRequest.getData());
+			Log.d("tutooo", response.toString());
```
Comment (FabianTerhorst) [inline library/src/main/java/com/pokegoapi/api/player/PlayerProfile.java]: Is the log needed on this place?

File: library/src/main/java/com/pokegoapi/api/PokemonGo.java (hunk 3 above)
Comment (iGio90) [inline library/src/main/java/com/pokegoapi/api/PokemonGo.java]: request type is needed!

File: library/src/main/java/com/pokegoapi/api/player/PlayerProfile.java (hunk 4 above)
Comment (iGio90) [inline library/src/main/java/com/pokegoapi/api/player/PlayerProfile.java]: No, these are my tests! I added it to check if the response was successfull and if you run it and filter for tutooo on logcat you should see the result. Once i got some score by you i'll cleanup the code. Note that im building it with terminal and without an IDE so if you see any other incorrect import, just tell me

File: library/src/main/java/com/pokegoapi/api/PokemonGo.java (hunk 3 above)
Comment (FabianTerhorst) [inline library/src/main/java/com/pokegoapi/api/PokemonGo.java]: can you say the line in the file where you used it.

Comment (iGio90) [inline library/src/main/java/com/pokegoapi/api/PokemonGo.java]: you are right, is now needed in the other class: PlayerProfile. Can be removed from there!

**Comment by iGio90** [comment]
@FabianTerhorst done! i left the log part to confirm the request.. please let me know! i'll remove it once its ready to merge (it is by my side)

File: library/src/main/java/com/pokegoapi/api/PokemonGo.java [hunk 5]
```diff
@@ -21,6 +21,7 @@
 
 import com.pokegoapi.api.device.DeviceInfo;
 import com.pokegoapi.api.device.DeviceInfos;
+
```
Comment (FabianTerhorst) [inline library/src/main/java/com/pokegoapi/api/PokemonGo.java]: can you try to remove all the changes from this file, because it doesn´t change to keep the filehistory clean.

**Comment by iGio90** [comment]
@FabianTerhorst here you are! im working on this without an IDE so tell me if there is something more to do!

File: library/src/main/java/com/pokegoapi/util/Signature.java [hunk 6]
```diff
@@ -79,8 +79,7 @@ public static void setSignature(PokemonGo api, RequestEnvelopeOuterClass.Request
 	}
 
 
-	private static int getLocationHash1(PokemonGo api, byte[] authTicket,
-										RequestEnvelopeOuterClass.RequestEnvelope.Builder builder) {
+	private static int getLocationHash1(PokemonGo api, byte[] authTicket, RequestEnvelopeOuterClass.RequestEnvelope.Builder builder) {
```
Comment (FabianTerhorst) [inline library/src/main/java/com/pokegoapi/util/Signature.java]: this change is also not related to this pr

File: library/src/main/java/com/pokegoapi/util/Signature.java [hunk 7]
```diff
@@ -115,4 +115,4 @@ private static long getRequestHash(byte[] authTicket, byte[] request) {
 		xx64.update(request, 0, request.length);
 		return xx64.getValue();
 	}
-}
+}
```
Comment (FabianTerhorst) [inline library/src/main/java/com/pokegoapi/util/Signature.java]: you removed the endline here

**Comment by iGio90** [comment]
@FabianTerhorst its quite impossible now to cleanup the things... i merged everything from here. Should i make a new PR?

**Comment by FabianTerhorst** [comment]
just add a new line after }. The merger can put them into a single commit.

**Comment by iGio90** [comment]
done! ty!

**Comment by Grover-c13** [comment]
Could you please do the two following things:
- make PokemonGo.getPlayer() lazy loaded in the PokemonGo object, (so only construct player object if it gets referenced once) to avoid the api making unnecessary API calls.
- Rebase the PR when you get a chance

**Comment by iGio90** [comment]
Goodmorning! I can, expect it in a couple of hour

**Comment by benjy3gg** [comment]
will you update this PR anytime soon?

**Comment by iGio90** [comment]
These code is perfectly working on me. The unique thing that could be changed is the static PlayerProfile that break the chance of using multiacct, so i set it to final on my local branch

**Comment by benjy3gg** [comment]
is it based on the latest Development branch?

**Comment by iGio90** [comment]
wait, im gonna merge and fix conflicts

**Comment by iGio90** [comment]
I will make a new PR in 10 mins with the latest things merged so it would be clean. Closing this!

**Comment by benjy3gg** [comment]
ty man!
